File: WorkoutTracker.Application/Exercises/ExercisePlan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WorkoutApplication.Domain.Exercises;
using WorkoutTracker.Application.Exercises;

namespace WorkoutApplication.Application.Exercises
{
    /// <summary>
    /// A class used to plan targets (reps and resistance) for an exercise prior to a workout.
    /// This class is not persisted, but rather used to gather information from the user and then 
    /// create the ExecutedWorkout instance.
    /// </summary>
    public class ExercisePlan
    {
        public int ExerciseInWorkoutId { get; set; }
        public int ExerciseId { get; set; }
        public string ExerciseName { get; set; }
        public byte NumberOfSets { get; set; }
        public SetType SetType { get; set; }
        public byte Sequence { get; set; }

        public byte TargetRepCountLastTime { get; set; }
        public byte MaxActualRepCountLastTime { get; set; }
        public byte? RecommendedTargetRepCount { get; set; }
        public byte TargetRepCount { get; set; }

        public decimal ResistanceAmountLastTime { get; set; }
        public string ResistanceMakeupLastTime { get; set; }
        public decimal? RecommendedResistanceAmount { get; set; }
        public string RecommendedResistanceMakeup { get; set; }
        public decimal ResistanceAmount { get; set; }
        public string ResistanceMakeup { get; set; }

        public string RecommendationReason { get; set; }

        public ExercisePlan(ExerciseInWorkout exercise)
        {
            ExerciseInWorkoutId = exercise.Id;
            ExerciseId = exercise.Exercise.Id;
            ExerciseName = exercise.Exercise.Name;
            NumberOfSets = exercise.NumberOfSets;
            SetType = exercise.SetType;
            Sequence = exercise.Sequence;
        }

        public void SetLastTimeValues(IEnumerable<ExecutedExercise> executedExercises)
        {
            TargetRepCountLastTime = executedExercises.Max(x => x.TargetRepCount);
            MaxActualRepCountLastTime = executedExercises.Max(x => x.ActualRepCount);

            var executedWithMaxResitance = 
                executedExercises.OrderByDescending(x => x.ResistanceAmount).First();

            ResistanceAmountLastTime = executedWithMaxResitance.ResistanceAmount;
            ResistanceMakeupLastTime = executedWithMaxResitance.ResistanceMakeup;
        }

        public void ApplyRecommendation(ExerciseAmountRecommendation recommendation)
        {
            if (recommendation != null)
            {
                RecommendationReason = recommendation.Reason;
                RecommendedResistanceAmount = recommendation.ResistanceAmount;
                RecommendedResistanceMakeup = recommendation.ResistanceMakeup;
                RecommendedTargetRepCount = recommendation.Reps;
            }
        }
    }
}

File: workout-tracker/src/app/workouts/_models/exercise-plan.ts
import { ResistanceType } from "../workout/_enums/resistance-type";
import { SetType } from "../workout/_enums/set-type";

export class ExercisePlan {
  public exerciseInWorkoutId: number = 0;
  public exerciseId: number = 0;
  public exerciseName: string = '';

  public numberOfSets: number = 0;
  public setType: SetType = SetType.Repetition;
  public resistanceType: ResistanceType = ResistanceType.BodyWeight;
  public sequence: number = 0;

  public targetRepCountLastTime: number = 0;
  public maxActualRepCountLastTime: number = 0;
  public avgActualRepCountLastTime: number = 0;

  public maxRangeOfMotionLastTime: number = 0;
  public avgRangeOfMotionLastTime: number = 0;

  public maxFormLastTime: number = 0;
  public avgFormLastTime: number = 0;

  public recommendedTargetRepCount: number | null = null;
  public targetRepCount: number | null = null; //Nullable because not all exercises use reps (example: running on a treadmill)

  public resistanceAmountLastTime: number = 0;
  public resistanceMakeupLastTime: string = '';

  public recommendedResistanceAmount: number | null = null;
  public recommendedResistanceMakeup: string | null = null;

  public resistanceAmount: number = 0;
  public resistanceMakeup: string | null = null; //Only applies to resistance band exercises

  public bandsEndToEnd: boolean | null = null;
  public involvesReps: boolean = true;
  public usesBilateralResistance: boolean = false;

  public recommendationReason: string = '';
}
